src/libexpr/nixexpr-as-json.cc

```diff
@@ -0,0 +1,326 @@
+#include "nixexpr.hh"
+#include "derivations.hh"
+#include "util.hh"
+
+#include <cstdlib>
+
+namespace nix {
+
+// binary operators are implemented in nixexpr.hh MakeBinOp
+
+// https://stackoverflow.com/questions/7724448
+// note: we use full jump table to make this as fast as possible
+// note: we assume valid input. errors should be handled by the nix parser
+// 93 * 7 = 651 byte
+char String_showAsJson_replace_array[93][7] = {
+  "\\u0000", "\\u0001", "\\u0002", "\\u0003", "\\u0004", // 0 - 4
+  "\\u0005", "\\u0006", "\\u0007", "\\b", "\\t", // 5 - 9
+  "\\n", "\\u000b", "\\f", "\\r", "\\u000e", // 10 - 14
+  "\\u000f", "\\u0010", "\\u0011", "\\u0012", "\\u0013", // 15 - 19
+  "\\u0014", "\\u0015", "\\u0016", "\\u0017", "\\u0018", // 20 - 24
+  "\\u0019", "\\u001a", "\\u001b", "\\u001c", "\\u001d", // 25 - 29
+  "\\u001e", "\\u001f", " ", "!", "\\\"", // 30 - 34
+  "#", "$", "%", "&", "'", // 35 - 39
+  "(", ")", "*", "+", ",", // 40 - 44
+  "-", ".", "/", "0", "1", // 45 - 49
+  "2", "3", "4", "5", "6", // 50 - 54
+  "7", "8", "9", ":", ";", // 55 - 59
+  "<", "=", ">", "?", "@", // 60 - 64
+  "A", "B", "C", "D", "E", // 65 - 69
+  "F", "G", "H", "I", "J", // 70 - 74
+  "K", "L", "M", "N", "O", // 75 - 79
+  "P", "Q", "R", "S", "T", // 80 - 84
+  "U", "V", "W", "X", "Y", // 85 - 89
+  "Z", "[", "\\\\", // 90 - 92
+};
+
+void String_showAsJson(std::ostream & o, const std::string & s) {
+  for (auto c = s.cbegin(); c != s.cend(); c++) {
+    if ((std::uint8_t) *c <= 92)
+      o << String_showAsJson_replace_array[(std::uint8_t) *c];
+    else
+      o << *c;
+  }
+}
+
+void Expr::showAsJson(std::ostream & str) const
+{
+    abort();
```

Comment:
This should probably produce a descriptive error where someone has to look to fix the situation. I could imagine that with the next subclass of Expr it might be easily forgotten and then this will only yield an obscure runtime error.

Reply:
uhm ... this is just copy-paste from `Expr::show` which i renamed to `Expr::showAsAterm`

Reply:
ugh, this sin't "end of input" but instead terminates the process.

Reply:
https://en.cppreference.com/w/cpp/utility/program/abort

Reply:
was added in [d4f0b0fc6c](https://github.com/NixOS/nix/commit/d4f0b0fc6cdb9ae2fd5fc057d621bc80b36a7b18#diff-029ba0c36fadf8ae76fc565815d5d197104ff4b60643ca945ce145eaadf7b2d7) by @edolstra - maybe he knows : )
my second guess is: this means "empty input"